File: util/resourcegrouptag/resource_group_tag.go
// Copyright 2021 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package resourcegrouptag

import (
	"github.com/pingcap/errors"
	"github.com/pingcap/kvproto/pkg/coprocessor"
	"github.com/pingcap/kvproto/pkg/kvrpcpb"
	"github.com/pingcap/tidb/parser"
	"github.com/pingcap/tidb/tablecodec/rowindexcodec"
	"github.com/pingcap/tipb/go-tipb"
	"github.com/tikv/client-go/v2/tikvrpc"
)

// EncodeResourceGroupTag encodes sql digest and plan digest into resource group tag.
func EncodeResourceGroupTag(sqlDigest, planDigest *parser.Digest, label tipb.ResourceGroupTagLabel) []byte {
	if sqlDigest == nil && planDigest == nil {
		return nil
	}

	tag := &tipb.ResourceGroupTag{Label: &label}
	if sqlDigest != nil {
		tag.SqlDigest = sqlDigest.Bytes()
	}
	if planDigest != nil {
		tag.PlanDigest = planDigest.Bytes()
	}
	b, err := tag.Marshal()
	if err != nil {
		return nil
	}
	return b
}

// DecodeResourceGroupTag decodes a resource group tag and return the sql digest.
func DecodeResourceGroupTag(data []byte) (sqlDigest []byte, err error) {
	if len(data) == 0 {
		return nil, nil
	}
	tag := &tipb.ResourceGroupTag{}
	err = tag.Unmarshal(data)
	if err != nil {
		return nil, errors.Errorf("invalid resource group tag data %x", data)
	}
	return tag.SqlDigest, nil
}

// GetResourceGroupLabelByKey determines the tipb.ResourceGroupTagLabel of key.
func GetResourceGroupLabelByKey(key []byte) tipb.ResourceGroupTagLabel {
	switch rowindexcodec.GetKeyKind(key) {
	case rowindexcodec.KeyKindRow:
		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelRow
	case rowindexcodec.KeyKindIndex:
		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelIndex
	default:
		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelUnknown
	}
}

// GetFirstKeyFromRequest gets the first Key of the request from tikvrpc.Request.
func GetFirstKeyFromRequest(req *tikvrpc.Request) (firstKey []byte) {
	if req == nil {
		return
	}
	switch req.Req.(type) {
	case *kvrpcpb.GetRequest:
		r := req.Req.(*kvrpcpb.GetRequest)
		if r != nil {
			firstKey = r.Key
		}
	case *kvrpcpb.BatchGetRequest:
		r := req.Req.(*kvrpcpb.BatchGetRequest)
		if r != nil && len(r.Keys) > 0 {
			firstKey = r.Keys[0]
		}
	case *kvrpcpb.ScanRequest:
		r := req.Req.(*kvrpcpb.ScanRequest)
		if r != nil {
			firstKey = r.StartKey
		}
	case *kvrpcpb.PrewriteRequest:
		r := req.Req.(*kvrpcpb.PrewriteRequest)
		if r != nil && len(r.Mutations) > 0 {
			if mutation := r.Mutations[0]; mutation != nil {
				firstKey = mutation.Key
			}
		}
	case *kvrpcpb.CommitRequest:
		r := req.Req.(*kvrpcpb.CommitRequest)
		if r != nil && len(r.Keys) > 0 {
			firstKey = r.Keys[0]
		}
	case *kvrpcpb.BatchRollbackRequest:
		r := req.Req.(*kvrpcpb.BatchRollbackRequest)
		if r != nil && len(r.Keys) > 0 {
			firstKey = r.Keys[0]
		}
	case *coprocessor.Request:
		r := req.Req.(*coprocessor.Request)
		if r != nil && len(r.Ranges) > 0 {
			if keyRange := r.Ranges[0]; keyRange != nil {
				firstKey = keyRange.Start
			}
		}
	case *coprocessor.BatchRequest:
		r := req.Req.(*coprocessor.BatchRequest)
		if r != nil && len(r.Regions) > 0 {
			if region := r.Regions[0]; region != nil {
				if len(region.Ranges) > 0 {
					if keyRange := region.Ranges[0]; keyRange != nil {
						firstKey = keyRange.Start
					}
				}
			}
		}
	}
	return
}

File: rust/crates/tidb-txnkv/src/resource_group_tag.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Resource-group tag decoding and request-key extraction from
//! `pkg/util/resourcegrouptag`.

use prost::Message;
use tidb_codec::{get_key_kind, KeyKind};
use tidb_proto::{
    CoprocessorBatchRequest, CoprocessorRequest, KvrpcBatchGetRequest, KvrpcBatchRollbackRequest,
    KvrpcCommitRequest, KvrpcGetRequest, KvrpcPessimisticLockRequest, KvrpcPrewriteRequest,
    KvrpcScanRequest, ResourceGroupTag, ResourceGroupTagLabel,
};

/// A malformed protobuf resource-group tag.
#[derive(Debug)]
pub struct ResourceGroupTagDecodeError {
    data: Vec<u8>,
    source: prost::DecodeError,
}

impl ResourceGroupTagDecodeError {
    /// Returns the malformed wire bytes.
    #[must_use]
    pub fn data(&self) -> &[u8] {
        &self.data
    }
}

impl std::fmt::Display for ResourceGroupTagDecodeError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str("invalid resource group tag data ")?;
        for byte in &self.data {
            write!(formatter, "{byte:02x}")?;
        }
        Ok(())
    }
}

impl std::error::Error for ResourceGroupTagDecodeError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        Some(&self.source)
    }
}

/// Decodes a resource-group tag and returns its SQL digest.
pub fn decode_resource_group_tag(
    data: &[u8],
) -> Result<Option<Vec<u8>>, ResourceGroupTagDecodeError> {
    if data.is_empty() {
        return Ok(None);
    }
    ResourceGroupTag::decode(data)
        .map(|tag| tag.sql_digest)
        .map_err(|source| ResourceGroupTagDecodeError {
            data: data.to_vec(),
            source,
        })
}

/// Classifies a key for resource-group attribution.
#[must_use]
pub fn get_resource_group_label_by_key(key: &[u8]) -> ResourceGroupTagLabel {
    match get_key_kind(key) {
        KeyKind::Row => ResourceGroupTagLabel::Row,
        KeyKind::Index => ResourceGroupTagLabel::Index,
        KeyKind::Unknown => ResourceGroupTagLabel::Unknown,
    }
}

/// Returns the first key carried by a request, or an empty slice when absent.
pub trait FirstKeyRequest {
    /// Returns the request's first key.
    fn first_key(&self) -> &[u8];
}

/// Gets a request's first key, returning an empty slice for no request.
#[must_use]
pub fn get_first_key_from_request<R: FirstKeyRequest + ?Sized>(request: Option<&R>) -> &[u8] {
    request.map_or(&[], FirstKeyRequest::first_key)
}

impl FirstKeyRequest for KvrpcGetRequest {
    fn first_key(&self) -> &[u8] {
        &self.key
    }
}

impl FirstKeyRequest for KvrpcBatchGetRequest {
    fn first_key(&self) -> &[u8] {
        self.keys.first().map_or(&[], Vec::as_slice)
    }
}

impl FirstKeyRequest for KvrpcScanRequest {
    fn first_key(&self) -> &[u8] {
        &self.start_key
    }
}

impl FirstKeyRequest for KvrpcPrewriteRequest {
    fn first_key(&self) -> &[u8] {
        self.mutations.first().map_or(&[], |mutation| &mutation.key)
    }
}

impl FirstKeyRequest for KvrpcCommitRequest {
    fn first_key(&self) -> &[u8] {
        self.keys.first().map_or(&[], Vec::as_slice)
    }
}

impl FirstKeyRequest for KvrpcBatchRollbackRequest {
    fn first_key(&self) -> &[u8] {
        self.keys.first().map_or(&[], Vec::as_slice)
    }
}

impl FirstKeyRequest for CoprocessorRequest {
    fn first_key(&self) -> &[u8] {
        self.ranges.first().map_or(&[], |range| &range.start)
    }
}

impl FirstKeyRequest for CoprocessorBatchRequest {
    fn first_key(&self) -> &[u8] {
        self.regions
            .first()
            .and_then(|region| region.ranges.first())
            .map_or(&[], |range| &range.start)
    }
}

impl FirstKeyRequest for KvrpcPessimisticLockRequest {
    fn first_key(&self) -> &[u8] {
        self.mutations.first().map_or(&[], |mutation| &mutation.key)
    }
}
